File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ParcelInfoReply.java
package com.lumiyaviewer.lumiya.slproto.messages;

import com.lumiyaviewer.lumiya.slproto.SLMessage;
import java.nio.ByteBuffer;
import java.util.UUID;

/**
 * ParcelInfoReply
 * dataserver -> simulator -> viewer
 * reliable
 *
 * <p>Template: {@code ParcelInfoReply Low 55 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code LLRemoteParcelInfoProcessor::processParcelInfoReply()} in indra/newview/llremoteparcelrequest.cpp
 * (secondlife/viewer @ c179f76c01).
 */
public class ParcelInfoReply extends SLMessage {
    public AgentData AgentData_Field;
    public Data Data_Field;

    /** Block AgentData, Single. */
    public static class AgentData {
        public UUID AgentID; // LLUUID
    }

    /** Block Data, Single. */
    public static class Data {
        public int ActualArea; // S32
        public int AuctionID; // S32
        public int BillableArea; // S32
        public byte[] Desc; // Variable 1
        public float Dwell; // F32
        public int Flags; // U8
        public float GlobalX; // F32 - meters
        public float GlobalY; // F32 - meters
        public float GlobalZ; // F32 - meters
        public byte[] Name; // Variable 1
        public UUID OwnerID; // LLUUID
        public UUID ParcelID; // LLUUID
        public int SalePrice; // S32
        public byte[] SimName; // Variable 1
        public UUID SnapshotID; // LLUUID
    }

    public ParcelInfoReply() {
        this.zeroCoded = true;
        this.AgentData_Field = new AgentData();
        this.Data_Field = new Data();
    }

    @Override
    public int CalcPayloadSize() {
        return this.Data_Field.Name.length + 33 + 1 + this.Data_Field.Desc.length + 4 + 4 + 1 + 4 + 4 + 4 + 1 + this.Data_Field.SimName.length + 16 + 4 + 4 + 4 + 20;
    }

    @Override
    public void Handle(SLMessageHandler messageHandler) {
        messageHandler.HandleParcelInfoReply(this);
    }

    @Override
    public void PackPayload(ByteBuffer byteBuffer) {
        // Message number: Low 55 (ParcelInfoReply).
        byteBuffer.putShort((short) 0xFFFF);
        byteBuffer.put((byte) 0x00);
        byteBuffer.put((byte) 0x37);
        packUUID(byteBuffer, this.AgentData_Field.AgentID);
        packUUID(byteBuffer, this.Data_Field.ParcelID);
        packUUID(byteBuffer, this.Data_Field.OwnerID);
        packVariable(byteBuffer, this.Data_Field.Name, 1);
        packVariable(byteBuffer, this.Data_Field.Desc, 1);
        packInt(byteBuffer, this.Data_Field.ActualArea);
        packInt(byteBuffer, this.Data_Field.BillableArea);
        packByte(byteBuffer, (byte) this.Data_Field.Flags);
        packFloat(byteBuffer, this.Data_Field.GlobalX);
        packFloat(byteBuffer, this.Data_Field.GlobalY);
        packFloat(byteBuffer, this.Data_Field.GlobalZ);
        packVariable(byteBuffer, this.Data_Field.SimName, 1);
        packUUID(byteBuffer, this.Data_Field.SnapshotID);
        packFloat(byteBuffer, this.Data_Field.Dwell);
        packInt(byteBuffer, this.Data_Field.SalePrice);
        packInt(byteBuffer, this.Data_Field.AuctionID);
    }

    @Override
    public void UnpackPayload(ByteBuffer byteBuffer) {
        this.AgentData_Field.AgentID = unpackUUID(byteBuffer);
        this.Data_Field.ParcelID = unpackUUID(byteBuffer);
        this.Data_Field.OwnerID = unpackUUID(byteBuffer);
        this.Data_Field.Name = unpackVariable(byteBuffer, 1);
        this.Data_Field.Desc = unpackVariable(byteBuffer, 1);
        this.Data_Field.ActualArea = unpackInt(byteBuffer);
        this.Data_Field.BillableArea = unpackInt(byteBuffer);
        this.Data_Field.Flags = unpackByte(byteBuffer) & 0xFF;
        this.Data_Field.GlobalX = unpackFloat(byteBuffer);
        this.Data_Field.GlobalY = unpackFloat(byteBuffer);
        this.Data_Field.GlobalZ = unpackFloat(byteBuffer);
        this.Data_Field.SimName = unpackVariable(byteBuffer, 1);
        this.Data_Field.SnapshotID = unpackUUID(byteBuffer);
        this.Data_Field.Dwell = unpackFloat(byteBuffer);
        this.Data_Field.SalePrice = unpackInt(byteBuffer);
        this.Data_Field.AuctionID = unpackInt(byteBuffer);
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/slproto/messages/ParcelInfoReply.kt
package com.lumiyaviewer.lumiya.slproto.messages

import com.lumiyaviewer.lumiya.slproto.SLMessage
import java.nio.ByteBuffer
import java.util.UUID

/**
 * ParcelInfoReply
 * dataserver -> simulator -> viewer
 * reliable
 *
 * <p>Template: {@code ParcelInfoReply Low 55 Trusted Zerocoded}
 * (recovered/reference/message_template.msg).
 * <p>Viewer reference: {@code LLRemoteParcelInfoProcessor::processParcelInfoReply()} in indra/newview/llremoteparcelrequest.cpp
 * (secondlife/viewer @ c179f76c01).
 */
open class ParcelInfoReply : SLMessage() {
    @JvmField var AgentData_Field: AgentData = AgentData()
    @JvmField var Data_Field: Data = Data()

    /** Block AgentData, Single. */
    open class AgentData {
        @JvmField var AgentID: UUID? = null
    }

    /** Block Data, Single. */
    open class Data {
        @JvmField var ActualArea: Int = 0
        @JvmField var AuctionID: Int = 0
        @JvmField var BillableArea: Int = 0
        @JvmField var Desc: ByteArray? = null
        @JvmField var Dwell: Float = 0f
        @JvmField var Flags: Int = 0
        @JvmField var GlobalX: Float = 0f
        @JvmField var GlobalY: Float = 0f
        @JvmField var GlobalZ: Float = 0f
        @JvmField var Name: ByteArray? = null
        @JvmField var OwnerID: UUID? = null
        @JvmField var ParcelID: UUID? = null
        @JvmField var SalePrice: Int = 0
        @JvmField var SimName: ByteArray? = null
        @JvmField var SnapshotID: UUID? = null
    }

    init {
        zeroCoded = true
    }

    override fun CalcPayloadSize(): Int {
        return Data_Field.Name!!.size + 33 + 1 + Data_Field.Desc!!.size + 4 + 4 + 1 + 4 + 4 + 4 + 1 + Data_Field.SimName!!.size + 16 + 4 + 4 + 4 + 20
    }

    override fun Handle(messageHandler: SLMessageHandler) {
        messageHandler.HandleParcelInfoReply(this)
    }

    override fun PackPayload(byteBuffer: ByteBuffer) {
        // Message number: Low 55 (ParcelInfoReply).
        byteBuffer.putShort(0xFFFF.toShort())
        byteBuffer.put((0x00).toByte())
        byteBuffer.put((0x37).toByte())
        packUUID(byteBuffer, AgentData_Field.AgentID)
        packUUID(byteBuffer, Data_Field.ParcelID)
        packUUID(byteBuffer, Data_Field.OwnerID)
        packVariable(byteBuffer, Data_Field.Name, 1)
        packVariable(byteBuffer, Data_Field.Desc, 1)
        packInt(byteBuffer, Data_Field.ActualArea)
        packInt(byteBuffer, Data_Field.BillableArea)
        packByte(byteBuffer, (Data_Field.Flags).toByte())
        packFloat(byteBuffer, Data_Field.GlobalX)
        packFloat(byteBuffer, Data_Field.GlobalY)
        packFloat(byteBuffer, Data_Field.GlobalZ)
        packVariable(byteBuffer, Data_Field.SimName, 1)
        packUUID(byteBuffer, Data_Field.SnapshotID)
        packFloat(byteBuffer, Data_Field.Dwell)
        packInt(byteBuffer, Data_Field.SalePrice)
        packInt(byteBuffer, Data_Field.AuctionID)
    }

    override fun UnpackPayload(byteBuffer: ByteBuffer) {
        AgentData_Field.AgentID = unpackUUID(byteBuffer)
        Data_Field.ParcelID = unpackUUID(byteBuffer)
        Data_Field.OwnerID = unpackUUID(byteBuffer)
        Data_Field.Name = unpackVariable(byteBuffer, 1)
        Data_Field.Desc = unpackVariable(byteBuffer, 1)
        Data_Field.ActualArea = unpackInt(byteBuffer)
        Data_Field.BillableArea = unpackInt(byteBuffer)
        Data_Field.Flags = unpackByte(byteBuffer).toInt() and 0xFF
        Data_Field.GlobalX = unpackFloat(byteBuffer)
        Data_Field.GlobalY = unpackFloat(byteBuffer)
        Data_Field.GlobalZ = unpackFloat(byteBuffer)
        Data_Field.SimName = unpackVariable(byteBuffer, 1)
        Data_Field.SnapshotID = unpackUUID(byteBuffer)
        Data_Field.Dwell = unpackFloat(byteBuffer)
        Data_Field.SalePrice = unpackInt(byteBuffer)
        Data_Field.AuctionID = unpackInt(byteBuffer)
    }
}
